libraries/rush-lib/src/logic/operations/LogChunksWritable.ts

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { TerminalWritable, type ITerminalChunk } from '@rushstack/terminal';
+
+import type { RushConfigurationProject } from '../../api/RushConfigurationProject';
+import { RushConstants } from '../RushConstants';
+import { getRelativeLogFilePathBase } from './ProjectLogWritable';
+
+import Stringer from 'stream-json/jsonl/Stringer';
+import { chain } from 'stream-chain';
+import type Chain from 'stream-chain';
+import * as fs from 'fs';
+
+/**
+ * A new terminal stream that writes all log chunks to a JSON format so they can be faithfully reconstructed
+ *  during build cache restores. This is used for adding warning + error messages in cobuilds where the original
+ *  logs cannot be completely restored from the existing `all.log` and `error.log` files.
+ */
+export class LogChunksWritable extends TerminalWritable {
+  public readonly logChunksPath: string;
+  public readonly relativeLogChunksPath: string;
+
+  private readonly _writeChain: Chain;
+
+  private readonly _chunks: ITerminalChunk[] = [];
+  private readonly _chunkFileStream: fs.WriteStream;
+
+  public constructor(
+    project: RushConfigurationProject,
+    /**
+     * A unique identifier for the log file. This is used to generate the prefix for the log file name.
+     */
+    logFilenameIdentifier: string
+  ) {
+    super();
+
+    const { logChunksPath, relativeLogChunksPath } = LogChunksWritable.getLogFilePaths({
+      project,
+      logFilenameIdentifier
+    });
+    this.logChunksPath = logChunksPath;
+    this.relativeLogChunksPath = relativeLogChunksPath;
+
+    this._chunkFileStream = fs.createWriteStream(logChunksPath);
+
+    this._writeChain = chain([new Stringer(), this._chunkFileStream]);
+  }
+
+  public static getLogFilePaths({
+    project,
+    logFilenameIdentifier,
+    isLegacyLog = false
+  }: {
+    project: RushConfigurationProject;
+    logFilenameIdentifier: string;
+    isLegacyLog?: boolean;
+  }): {
+    logChunksPath: string;
+    relativeLogChunksPath: string;
+  } {
+    const { projectFolder } = project;
+
+    const logFileBaseName: string = getRelativeLogFilePathBase(
+      project,
+      logFilenameIdentifier,
+      isLegacyLog,
+      `.rush/${RushConstants.rushTempFolderName}/operations`
+    );
+    const relativeLogChunksPath: string = `${logFileBaseName}.chunks.jsonl`;
```

Comment:
What is `.jsonl`?  Should we be using `.json` file extension?

Reply:
I see,  https://jsonlines.org/ is a somewhat standard format where JSON objects are delimited by newlines.  Example output:

**libraries/tree-pattern/.rush/temp/operations/tree-pattern._phase_build.chunks.jsonl**
```jsonl
{"text":"Invoking: heft run --only build -- --clean \r\n","kind":"O"}
{"text":" ---- lifecycle started ---- \r\n","kind":"O"}
{"text":" ---- build started ---- \r\n","kind":"O"}
{"text":"[build:clean] Deleted 0 files and 4 folders\r\n","kind":"O"}
{"text":"[build:typescript] Using TypeScript version 5.0.4\r\n","kind":"O"}
{"text":"[build:lint] Using ESLint version 7.30.0\r\n","kind":"O"}
{"text":"[build:api-extractor] Using API Extractor version 7.35.3\r\n","kind":"O"}
{"text":"[build:api-extractor] Analysis will use the bundled TypeScript version 5.0.4\r\n","kind":"O"}
{"text":" ---- build finished (4.507s) ---- \r\n","kind":"O"}
{"text":" ---- lifecycle finished (4.509s) ---- \r\n","kind":"O"}
{"text":"-------------------- Finished (4.511s) --------------------\r\n","kind":"O"}
```

Let's explain this in a code comment with an example filename and output.